src/couch_jobs/test/couch_jobs_tests.erl

```diff
@@ -0,0 +1,607 @@
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_jobs_tests).
+
+
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("couch/include/couch_eunit.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+
+% Job creation API can take an undefined Tx object
+% in that case it will start its own transaction
+-define(TX, undefined).
+
+
+couch_jobs_basic_test_() ->
+    {
+        "Test couch jobs basics",
+        {
+            setup,
+            fun setup_couch/0, fun teardown_couch/1,
+            {
+                foreach,
+                fun setup/0, fun teardown/1,
+                [
+                    fun add_remove_pending/1,
+                    fun add_remove_errors/1,
+                    fun get_job_data_and_state/1,
+                    fun resubmit_as_job_creator/1,
+                    fun type_timeouts_and_server/1,
+                    fun dead_notifier_restarts_jobs_server/1,
+                    fun bad_messages_restart_couch_jobs_server/1,
+                    fun bad_messages_restart_notifier/1,
+                    fun bad_messages_restart_activity_monitor/1,
+                    fun basic_accept_and_finish/1,
+                    fun accept_blocking/1,
+                    fun job_processor_update/1,
+                    fun resubmit_enqueues_job/1,
+                    fun add_with_scheduled_time/1,
+                    fun resubmit_custom_schedtime/1,
+                    fun accept_max_schedtime/1,
+                    fun accept_random_ids/1,
+                    fun subscribe/1,
+                    fun subscribe_wait_multiple/1,
+                    fun enqueue_inactive/1,
+                    fun remove_running_job/1,
+                    fun check_get_jobs/1,
+                    fun use_fabric_transaction_object/1
+                ]
+            }
+        }
+    }.
+
+
+setup_couch() ->
+    test_util:start_couch([fabric]).
+
+
+teardown_couch(Ctx) ->
+    test_util:stop_couch(Ctx),
+    meck:unload().
+
+
+setup() ->
+    couch_jobs_fdb:clear_jobs(),
+    application:start(couch_jobs),
+    T1 = {<<"t1">>, 1024}, % a complex type should work
+    T2 = 42, % a number should work as well
+    T1Timeout = 2,
+    T2Timeout = 3,
+    couch_jobs:set_type_timeout(T1, T1Timeout),
+    couch_jobs:set_type_timeout(T2, T2Timeout),
+    #{
+        t1 => T1,
+        t2 => T2,
+        t1_timeout => T1Timeout,
+        j1 => <<"j1">>,
+        j2 => <<"j2">>,
+        dbname => ?tempdb()
+    }.
+
+
+teardown(#{dbname := DbName}) ->
+    application:stop(couch_jobs),
+    couch_jobs_fdb:clear_jobs(),
```

Comment:
I wonder if we shouldn't move debug helpers that make modifications to a module in the test directory so they're not deployed? Read-only debug would be helpful in production but I'd worry a bit about having things that could wipe out all active jobs.

Reply:
Good call